internal/core/app_deploy.go

```diff
@@ -313,14 +315,48 @@ func (op *deployOperation) Do(ctx context.Context, log hclog.Logger, app *App, m
 		return nil, err
 	}
 
+	// NOTE(izaak): We do this to satisfy argmamper - because the DeployFunc asks for this,
+	// we need to prove to argmapper that we can supply one. We'll swap this out
+	// with a different pointer on the plugin side. This one doesn't end up being used for anything.
+	declaredResourcesResp := &component.DeclaredResourcesResp{}
+
+	var args []argmapper.Arg
+	args = append(args, argmapper.Typed(declaredResourcesResp))
+	args = append(args, op.args()...)
+
 	val, err := app.callDynamicFunc(ctx,
 		log,
 		(*component.Deployment)(nil),
 		op.component,
 		op.component.Value.(component.Platform).DeployFunc(),
-		op.args()...,
+		args...,
 	)
 
+	// NOTE(izaak): Unfortunately, the returned type here is an waypoint-plugin-sdk/internal/plugioncomponent/Deployment,
+	// and because it's internal we can't cast to it. This is similar to how we're getting TemplateData
+	// (https://github.com/hashicorp/waypoint-plugin-sdk/blob/eda7ae600c2d69ba042ecac4d6d9333d37f6fd9d/docs/func.go#L106),
+	// but it still doesn't feel great.
+	var pluginDeclaredResources []*sdkproto.DeclaredResource
+	pluginDeclaredResourcesField := reflect.ValueOf(val).Elem().FieldByName("DeclaredResources")
```

Comment:
You mention doing this like TemplateData, we should probably mask the digging around in a reflect value the same as we do with TemplateData.

Reply:
No longer needed! It gave me the willies too.